src/connection.rs

```diff
@@ -485,16 +486,24 @@ where
                     Ok(transport) => transport,
                     Err(e) => {
                         warn!(%e, "Failed to connect to broker");
+                        errors.push(e);
                         continue;
                     }
                 };
 
                 return ControlFlow::Break(connection);
             }
 
-            let err = Box::<dyn std::error::Error + Send + Sync>::from(
-                "Failed to connect to any broker, backing off".to_string(),
-            );
+            // errors should always contain at least 1 element here.
+            let errors_string = errors
+                .into_iter()
+                .map(|e| e.to_string())
```

Comment:
I suggest you create a helper type here to delay the error formatting until it is actually needed:

```rust
struct MultiError(Vec<Box<dyn std::error::Error + Send + Sync>);

impl std::error::Error for MultiError {...}
```

Reply:
Sure, that's probably better. Maybe it would even be more useful with


```
struct MultiError<T: std::error::Error + Send + Sync + 'static>(Vec<T>)
```

Reply:
Though since we're sticking it in a `Box<dyn ...>` anyway I guess it doesn't make much of a difference in this case.